File: common-sdk/src/main/java/ru/usedesk/common_sdk/internal/UsedeskFileUtil.java
package ru.usedesk.common_sdk.internal;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.OpenableColumns;
import android.webkit.MimeTypeMap;

public class UsedeskFileUtil {
    public static String getFileName(Context context, Uri uri) {
        return getFileName(context.getContentResolver(), uri);
    }

    public static String getFileName(ContentResolver contentResolver, Uri uri) {
        String result = null;
        String schema = uri.getScheme();
        if (schema != null && schema.equals("content")) {
            try (Cursor cursor = contentResolver.query(uri,
                    null,
                    null,
                    null,
                    null)) {
                if (cursor != null && cursor.moveToFirst()) {
                    result = cursor.getString(cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME));
                }
            }
        }
        if (result == null) {
            String path = uri.getPath();
            if (path != null) {
                int cut = path.lastIndexOf('/');
                if (cut >= 0) {
                    result = path.substring(cut + 1);
                } else {
                    result = path;
                }
            }
        }
        if (result != null) {
            return result;
        } else {
            return "";
        }
    }

    public static String getMimeType(Context context, Uri uri) {
        return getMimeType(context.getContentResolver(), uri);
    }


    public static String getMimeType(ContentResolver contentResolver, Uri uri) {
        String mimeType;
        if (ContentResolver.SCHEME_CONTENT.equals(uri.getScheme())) {
            mimeType = contentResolver.getType(uri);
        } else {
            String extension = MimeTypeMap.getFileExtensionFromUrl(uri.toString()).toLowerCase();
            mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
        }
        if (mimeType != null) {
            return mimeType;
        } else {
            return "";
        }
    }
}

File: common-sdk/src/main/java/ru/usedesk/common_sdk/utils/UsedeskFileUtil.kt

package ru.usedesk.common_sdk.utils

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import android.webkit.MimeTypeMap

object UsedeskFileUtil {
    fun Context.getFileName(uri: Uri) = contentResolver.getFileName(uri)

    fun Context.getFileSize(uri: Uri) = contentResolver.getFileSize(uri)

    fun Context.getMimeType(uri: Uri) = contentResolver.getMimeType(uri)

    fun ContentResolver.getFileSize(uri: Uri): Long {
        var size = -1L
        try {
            when (uri.scheme) {
                ContentResolver.SCHEME_CONTENT -> query(
                    uri,
                    null,
                    null,
                    null,
                    null
                )?.use { cursor ->
                    if (cursor.moveToFirst()) {
                        val columnIndex = cursor.getColumnIndexOrThrow(OpenableColumns.SIZE)
                        size = cursor.getLong(columnIndex)
                    }
                }
                ContentResolver.SCHEME_FILE -> openAssetFileDescriptor(
                    uri,
                    "r"
                )?.use {
                    size = it.length
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return size
    }

    fun ContentResolver.getFileName(uri: Uri): String {
        var name: String? = null
        if (uri.scheme == ContentResolver.SCHEME_CONTENT) {
            try {
                query(
                    uri,
                    null,
                    null,
                    null,
                    null
                )?.use { cursor ->
                    if (cursor.moveToFirst()) {
                        val index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                        name = cursor.getString(index)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        if (name == null) {
            try {
                uri.path?.also {
                    val cut = it.lastIndexOf('/')
                    name = when {
                        cut >= 0 -> it.substring(cut + 1)
                        else -> it
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return name ?: ""
    }

    fun ContentResolver.getMimeType(uri: Uri) = try {
        when (uri.scheme) {
            ContentResolver.SCHEME_CONTENT -> getType(uri)
            else -> {
                val extension = MimeTypeMap.getFileExtensionFromUrl(uri.toString()).lowercase()
                MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
            }
        }
    } catch (e: Exception) {
        e.printStackTrace()
        null
    } ?: ""
}